storage-proofs-update/src/circuit.rs

```diff
@@ -0,0 +1,590 @@
+            );
+        }
+        // Assert that `h_select` is valid. HSelect should be a uint whose binary representation has
+        // exactly 1 of its first `h_select_bit_len` (i.e. 6) bits set.
+        if let Some(h_select) = h_select {
+            let mut allowed_h_select_values = (0..h_select_bit_len).map(|i| Fr::from(1u64 << i));
+            assert!(allowed_h_select_values.any(|allowed_h_select| allowed_h_select == h_select));
+        }
+        assert_eq!(apex_leafs.len(), apex_leaf_count);
+        assert_eq!(partition_path.len(), partition_bit_len);
+        assert!(partition_path.iter().all(|siblings| siblings.len() == 1));
+        assert_eq!(challenge_proofs.len(), challenge_count);
+        // Check that all partition challenge's have the same same partition path.
+        for challenge_proof in &challenge_proofs[1..] {
+            assert_eq!(
+                &challenge_proof.path_d_new[challenge_bit_len - partition_bit_len..],
+                partition_path,
+            );
+        }
+
+        // Allocate public-inputs.
+
+        let partition = AllocatedNum::alloc(cs.namespace(|| "partition_index"), || {
+            k.ok_or(SynthesisError::AssignmentMissing)
+        })?;
+        partition.inputize(cs.namespace(|| "partition_index (public input)"))?;
+
+        let comm_r_old = AllocatedNum::alloc(cs.namespace(|| "comm_r_old"), || {
+            comm_r_old.ok_or(SynthesisError::AssignmentMissing)
+        })?;
+        comm_r_old.inputize(cs.namespace(|| "comm_r_old_input"))?;
+
+        let comm_d_new = AllocatedNum::alloc(cs.namespace(|| "comm_d_new"), || {
+            comm_d_new.ok_or(SynthesisError::AssignmentMissing)
+        })?;
+        comm_d_new.inputize(cs.namespace(|| "comm_d_new_input"))?;
+
+        let comm_r_new = AllocatedNum::alloc(cs.namespace(|| "comm_r_new"), || {
+            comm_r_new.ok_or(SynthesisError::AssignmentMissing)
+        })?;
+        comm_r_new.inputize(cs.namespace(|| "comm_r_new_input"))?;
+
+        let h_select = AllocatedNum::alloc(cs.namespace(|| "h_select"), || {
+            h_select.ok_or(SynthesisError::AssignmentMissing)
+        })?;
+        h_select.inputize(cs.namespace(|| "h_select_input"))?;
+
+        // Allocate values derived from public-inputs.
+
+        // Allocate the partition-index as bits.
+        let partition_bits: Vec<AllocatedBit> =
+            allocated_num_to_allocated_bits(cs.namespace(|| "partition_bits"), &partition)?
+                .into_iter()
+                .take(partition_bit_len)
+                .collect();
+
+        // Allocate the six least-significant bits of `h_select`.
+        let h_select_bits: Vec<AllocatedBit> =
+            allocated_num_to_allocated_bits(cs.namespace(|| "h_select_bits"), &h_select)?
+                .into_iter()
+                .take(h_select_bit_len)
+                .collect();
+
+        // phi = H(comm_d_new || comm_r_old)
+        let phi = <TreeR::Hasher as Hasher>::Function::hash2_circuit(
```

Comment:
Should this really depend on the treeR hasher, or should this just be fixed to be poseidon?

Reply:
I have hardcoded `phi` and `rho` to use Poseidon.